pkg/traceroute/icmp.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package traceroute
+
+import (
+	"bytes"
+	"encoding/binary"
+	"log"
+	"net"
+	"time"
+
+	"golang.org/x/net/ipv4"
+)
+
+func (t *Trace) SendTracesICMP4() {
+	conn, err := net.ListenPacket("ip4:icmp", t.src.String())
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer conn.Close()
+
+	rSocket, err := ipv4.NewRawConn(conn)
+	if err != nil {
+		log.Fatal("can not create raw socket:", err)
+	}
+	id := uint16(1)
+	mod := uint16(1 << 15)
+	for ttl := 1; ttl <= int(t.MaxHops); ttl++ {
+		for j := 0; j < t.TracesPerHop; j++ {
+			hdr, payload := t.BuildIPv4ICMP(uint8(ttl), id, id, 0)
+			rSocket.WriteTo(hdr, payload, nil)
+			pb := &Probe{
+				id:       uint32(hdr.ID),
+				dest:     [4]byte(t.Dest.To4()),
+				ttl:      ttl,
+				sendtime: time.Now(),
+			}
+			t.SendChan <- pb
+			id = (id + 1) % mod
+			go t.ReceiveTracesICMP4()
+			time.Sleep(time.Microsecond * time.Duration(100000/t.PacketRate))
+		}
+	}
+	for _, port := range []uint16{
+		80,
+		8080,
+		443,
+		8443,
+	} {
+		go t.IPv4TCPProbe(port)
+	}
+}
+
+func (t *Trace) ReceiveTracesICMP4() {
```

Comment:
pretty much anywhere you call log.Fatal -- would be best if it were an error return.

Reply:
agreed